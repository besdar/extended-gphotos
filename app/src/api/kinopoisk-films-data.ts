

import { ImageItemType } from "../Components/ImageItem/ImageItem";
import { FilmsDataType } from "./kinopoisk-films-object";

// https://kinopoiskapiunofficial.tech
const API_KEY = 'YOUR_API_KEY';

type KinopoiskFrameType = {
    image: string,
    preview: string,
}

type KinopoiskFrameResponseType = {
    frames: Array<KinopoiskFrameType>
}

const getFramesFromAPI = async (filmId: number): Promise<KinopoiskFrameResponseType> => {
    const response = await fetch(`https://kinopoiskapiunofficial.tech/api/v2.1/films/${filmId}/frames`, {
        headers: {
            'X-API-KEY': API_KEY
        }
    });

    if (response.status === 200) {
        return response.json()
    }

    return {
        frames: []
    };
};


type KinopoiskFilmInfoType = {
    nameRu: string,
    webUrl: string,
    // other staff
}

type KinopoiskFilmInfoResponseType = {
    data: KinopoiskFilmInfoType,
    // other staff
}

const getFilmInfoFromAPI = async (filmId: number): Promise<KinopoiskFilmInfoType | null> => {
    const response = await fetch(`https://kinopoiskapiunofficial.tech/api/v2.1/films/${filmId}`, {
        headers: {
            'X-API-KEY': API_KEY
        }
    });

    if (response.status === 200) {
        const { data: result } = await response.json() as KinopoiskFilmInfoResponseType;

        return Promise.resolve({
            nameRu: result.nameRu,
            webUrl: result.webUrl
        })
    }

    return null;
}

// Caution: This is very resource-intensive function
export const getFilmCovers = async (filmsIds: Array<number>) => (
    new Promise(resolve => {
        const result = [] as FilmsDataType;
        const filmsIdsLength = filmsIds.length;

        filmsIds.forEach((filmId, index) => {
            setTimeout(async () => {
                const framesObject = await getFramesFromAPI(filmId);
                const infoObject = await getFilmInfoFromAPI(filmId);

                if (infoObject !== null) {
                    const newElements = framesObject.frames.filter(Object.entries).map(el => ({
                        ...el,
                        ...infoObject,
                    }));

                    result.push(...newElements);
                }


                if (index + 1 === filmsIdsLength) {
                    resolve(result)
                }
            }, index * 1000); // prevent ban from server
        })
    })
);

export const getRandomKinopoiskImageInfo = (ImagesData: FilmsDataType) => (previewImage = true): ImageItemType => {
    const randomIndex = Math.floor(Math.random() * (ImagesData.length - 1));

    return {
        imageUrl: previewImage ? ImagesData[randomIndex].preview : ImagesData[randomIndex].image,
        tooltipTitle: ImagesData[randomIndex].nameRu,
        webUrl: ImagesData[randomIndex].webUrl,
    };
};