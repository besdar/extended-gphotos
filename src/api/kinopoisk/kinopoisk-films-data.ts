import { ImageItemType } from '../../Components/ImageItem/ImageItem';
import { getFilmInfoFromAPI, getFramesFromAPI } from './api';
import { FilmsDataType } from './kinopoisk-films-object';

// Caution: This is very resource-intensive function
export const getFilmCovers = async (filmsIds: Array<number>): Promise<FilmsDataType> => (
  new Promise((resolve) => {
    const result = [] as FilmsDataType;
    const filmsIdsLength = filmsIds.length;

    filmsIds.forEach((filmId, index) => {
      setTimeout(async () => {
        const framesObject = await getFramesFromAPI(filmId);
        const infoObject = await getFilmInfoFromAPI(filmId);
        const newElements = framesObject.frames.filter(Object.entries).map((el) => ({
          ...el,
          ...infoObject,
        }));

        if (newElements.length) {
          result.push(...newElements);
        }

        if (index + 1 === filmsIdsLength) {
          resolve(result);
        }
      }, index * 1000); // prevent ban from server
    });
  })
);

export const getRandomKinopoiskImageInfo = (ImagesData: FilmsDataType) => (
  (previewImage = true): ImageItemType => {
    const randomIndex = Math.floor(Math.random() * (ImagesData.length - 1));

    return {
      imageUrl: previewImage ? ImagesData[randomIndex].preview : ImagesData[randomIndex].image,
      tooltipTitle: ImagesData[randomIndex].nameRu,
      webUrl: ImagesData[randomIndex].webUrl,
    };
  }
);
