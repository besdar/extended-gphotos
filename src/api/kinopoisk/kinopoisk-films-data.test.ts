// import { getFilmInfoFromAPI, getFramesFromAPI } from "./api";
import { getRandomKinopoiskImageInfo } from './kinopoisk-films-data';
import { FilmInfo } from './kinopoisk-films-object';

// jest.mock('./api', () => ({
//     getFilmInfoFromAPI: jest.fn(),
//     getFramesFromAPI: jest.fn(),
// }));

describe('Tests for getRandomKinopoiskImageInfo function', () => {
  const ImageData = [
    {
      image: 'image1',
      nameRu: 'nameRu1',
      preview: 'prewiew1',
      webUrl: 'webUrl1',
    },
  ] as Array<FilmInfo>;

  it('Returns random object with preview image', () => {
    expect(getRandomKinopoiskImageInfo(ImageData)()).toMatchSnapshot();
  });

  it('Returns random object without preview image', () => {
    expect(getRandomKinopoiskImageInfo(ImageData)(false)).toMatchSnapshot();
  });
});

// TODO: Can not mock setTimeout with Promises. setTimeout never triggers,
// Promises (fetch) never resolves or return undefined
// describe('Tests for getFilmCovers function', () => {
//     beforeAll(() => {
//         jest.useFakeTimers('legacy');

//         (getFilmInfoFromAPI as jest.Mock).mockImplementation(async () => Promise.resolve({
//             data: { nameRu: 'nameRu', webUrl: 'webUrl' }
//         }));

//         (getFramesFromAPI as jest.Mock).mockImplementation(async () => Promise.resolve({
//             frames: [{ image: 'image', preview: 'preview' }]
//         }));
//     });

//     it('Create film data from api', () => {
//         const resultPromise = getFilmCovers([1, 2]);

//         jest.runAllTimers();

//         return resultPromise.then(result => {
//             expect(result).not.toHaveLength(0);
//             expect(result).toMatchSnapshot();
//         });
//     })
// });
