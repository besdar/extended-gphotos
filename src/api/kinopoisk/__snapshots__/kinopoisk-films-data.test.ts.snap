// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for getRandomKinopoiskImageInfo function Returns random object with preview image 1`] = `
Object {
  "imageUrl": "prewiew1",
  "tooltipTitle": "nameRu1",
  "webUrl": "webUrl1",
}
`;

exports[`Tests for getRandomKinopoiskImageInfo function Returns random object without preview image 1`] = `
Object {
  "imageUrl": "image1",
  "tooltipTitle": "nameRu1",
  "webUrl": "webUrl1",
}
`;
